File: tests/pdf_processor/utils/test_helpers.py
import pytest
from pdf_processor.utils.helpers import rect_to_dict

def test_rect_to_dict_none():
    """Test passing None returns None."""
    assert rect_to_dict(None) is None

def test_rect_to_dict_basic():
    """Test passing a valid 4-element list of coordinates."""
    rect = [10.5, 20.2, 110.5, 120.2]
    expected = {
        "x0": 10.5,
        "y0": 20.2,
        "x1": 110.5,
        "y1": 120.2,
        "width": 100.0,
        "height": 100.0,
    }
    assert rect_to_dict(rect) == expected

def test_rect_to_dict_with_page_height():
    """Test passing a valid rect and page_height adds canvas properties."""
    rect = [10.5, 20.2, 110.5, 120.2]
    page_height = 800.0
    expected = {
        "x0": 10.5,
        "y0": 20.2,
        "x1": 110.5,
        "y1": 120.2,
        "width": 100.0,
        "height": 100.0,
        "canvas_top": 679.8,     # 800.0 - 120.2
        "canvas_bottom": 779.8,  # 800.0 - 20.2
    }
    assert rect_to_dict(rect, page_height=page_height) == expected

def test_rect_to_dict_invalid_size():
    """Test passing a list with incorrect number of elements returns None."""
    # Too few
    assert rect_to_dict([10, 20, 100]) is None
    # Too many
    assert rect_to_dict([10, 20, 100, 200, 300]) is None

def test_rect_to_dict_invalid_elements():
    """Test passing non-convertible elements returns None."""
    assert rect_to_dict(["a", "b", "c", "d"]) is None
    assert rect_to_dict([10, 20, "c", 100]) is None

def test_rect_to_dict_invalid_type():
    """Test passing a completely invalid type returns None."""
    assert rect_to_dict("not a list or tuple") is None
    assert rect_to_dict({"x": 10, "y": 20}) is None

File: tests/pdf_processor/utils/test_helpers.js
const { rectToDict, getPageDimensions } = require('../../../src/pdf_processor/utils/helpers');

describe('helpers.js', () => {

    describe('getPageDimensions', () => {
        it('returns standard size when no mediaBox is provided', () => {
            const [width, height] = getPageDimensions(null);
            expect(width).toBe(595.27);
            expect(height).toBe(842.0);
        });

        it('calculates width and height from valid mediaBox', () => {
            const mockPageInfo = { mediaBox: [0, 0, 500, 600] };
            const [width, height] = getPageDimensions(mockPageInfo);
            expect(width).toBe(500.0);
            expect(height).toBe(600.0);
        });

        it('handles non-zero origin mediaBox', () => {
            const mockPageInfo = { mediaBox: [50, 60, 550, 660] };
            const [width, height] = getPageDimensions(mockPageInfo);
            expect(width).toBe(500.0);
            expect(height).toBe(600.0);
        });
    });

    describe('rectToDict', () => {
        it('returns null for empty or invalid array', () => {
            expect(rectToDict(null)).toBeNull();
            expect(rectToDict([10, 20, 100])).toBeNull();
            expect(rectToDict(["a", "b", "c", "d"])).toBeNull();
        });

        it('returns a basic dict of coordinates', () => {
            const rect = [10.5, 20.2, 110.5, 120.2];
            const expected = {
                x0: 10.5,
                y0: 20.2,
                x1: 110.5,
                y1: 120.2,
                width: 100.0,
                height: 100.0,
            };
            expect(rectToDict(rect)).toEqual(expected);
        });

        it('returns a dict with canvas properties when page height is provided', () => {
            const rect = [10.5, 20.2, 110.5, 120.2];
            const expected = {
                x0: 10.5,
                y0: 20.2,
                x1: 110.5,
                y1: 120.2,
                width: 100.0,
                height: 100.0,
                canvas_top: 679.8,
                canvas_bottom: 779.8,
            };
            expect(rectToDict(rect, 800.0)).toEqual(expected);
        });
    });

});
